sycl/include/sycl/usm.hpp

```diff
@@ -19,161 +19,316 @@ namespace sycl {
+#endif
+);
 __SYCL_EXPORT void *
 malloc(size_t size, const queue &q, usm::alloc kind,
-       const property_list &propList,
-       const detail::code_location CodeLoc = detail::code_location::current());
+       const property_list &propList
+#ifndef DISABLE_SYCL_INSTRUMENTATION_METADATA
+       ,
+       const detail::code_location &CL = detail::code_location::current()
+#endif
+);
 
-__SYCL_EXPORT void *aligned_alloc(
-    size_t alignment, size_t size, const device &dev, const context &ctxt,
-    usm::alloc kind,
-    const detail::code_location CodeLoc = detail::code_location::current());
-__SYCL_EXPORT void *aligned_alloc(
-    size_t alignment, size_t size, const device &dev, const context &ctxt,
-    usm::alloc kind, const property_list &propList,
-    const detail::code_location CodeLoc = detail::code_location::current());
-__SYCL_EXPORT void *aligned_alloc(
-    size_t alignment, size_t size, const queue &q, usm::alloc kind,
-    const detail::code_location CodeLoc = detail::code_location::current());
-__SYCL_EXPORT void *aligned_alloc(
-    size_t alignment, size_t size, const queue &q, usm::alloc kind,
-    const property_list &propList,
-    const detail::code_location CodeLoc = detail::code_location::current());
+__SYCL_EXPORT void *
+aligned_alloc(size_t alignment, size_t size, const device &dev,
+              const context &ctxt, usm::alloc kind
+#ifndef DISABLE_SYCL_INSTRUMENTATION_METADATA
+              ,
+              const detail::code_location &CL = detail::code_location::current()
+#endif
+);
+__SYCL_EXPORT void *
+aligned_alloc(size_t alignment, size_t size, const device &dev,
+              const context &ctxt, usm::alloc kind,
+              const property_list &propList
+#ifndef DISABLE_SYCL_INSTRUMENTATION_METADATA
+              ,
+              const detail::code_location &CL = detail::code_location::current()
+#endif
+);
+__SYCL_EXPORT void *
+aligned_alloc(size_t alignment, size_t size, const queue &q, usm::alloc kind
+#ifndef DISABLE_SYCL_INSTRUMENTATION_METADATA
+              ,
+              const detail::code_location &CL = detail::code_location::current()
+#endif
+);
+__SYCL_EXPORT void *
+aligned_alloc(size_t alignment, size_t size, const queue &q, usm::alloc kind,
+              const property_list &propList
+#ifndef DISABLE_SYCL_INSTRUMENTATION_METADATA
+              ,
+              const detail::code_location &CL = detail::code_location::current()
+#endif
+);
 
 ///
 // Template forms
 ///
 template <typename T>
 T *malloc_device(
     size_t Count, const device &Dev, const context &Ctxt,
-    const property_list &PropList = {},
-    const detail::code_location CodeLoc = detail::code_location::current()) {
+    const property_list &PropList = {}
+#ifndef DISABLE_SYCL_INSTRUMENTATION_METADATA
+    ,
+    const detail::code_location &CodeLoc = detail::code_location::current()
+#endif
+) {
+#ifdef DISABLE_SYCL_INSTRUMENTATION_METADATA
+  const detail::code_location &CodeLoc = {};
+#endif
```

Comment:
I do like C macros but not when they hide "," or require anything from following code, like you show.

Reply:
I've missed @tovinkere 's comment while writing my response, but I do think that his approach is way nicer and easier to maintain than current version. All those unnecessary preprocessor `if`s are so hard to read and they just steal attention from what's really meaningful in the code.

Reply:
> require anything from following code

We can avoid that by standardizing the name of the code location and declaring it as `inline constexpr` variable in the namespace scope.

Reply:
> I do like C macros but not when they hide "," or require anything from following code, like you show.

I am in general agreement with your position, but this use requires an #ifndef and #ifdef pair that makes it easy for someone to make a mistake that may go unnoticed. Using these macros is one way to prevent mistakes and have consistency.

Reply:
I've uploaded a patch that uses macros, which you'd hopefully like better.

Reply:
I think this PR should be split in two. The refactoring you've done for handling code location is rather significant.

Reply:
> I think this PR should be split in two. The refactoring you've done for handling code location is rather significant.

I've started with a simple change where properties were not dropped, but other reviewers requested a more complete change. At this point I do not see a benefit of splitting this PR. Just pretend that this refactoring is the goal of this PR although it also fixes some small functional issue.

Reply:
> > I think this PR should be split in two. The refactoring you've done for handling code location is rather significant.
> 
> I've started with a simple change where properties were not dropped, but other reviewers requested a more complete change. At this point I do not see a benefit of splitting this PR. Just pretend that this refactoring is the goal of this PR although it also fixes some small functional issue.

I am okay with not having two PRs for this - the one thing I do not like is that queue.hpp and usm.hpp implement two different sets of macros. I would hate to have a PR just to reconcile the different approaches.

Reply:
These aren't functional macros but rather some convenience ones (and quite simple) that some **may** prefer using over more explicit coding. I can imagine (and support) people creating a third variant if they like it better. Whose final call is it how to proceed with this PR?

@steffenlarsen : I've change the title to reflect it is an ABI breaking change, but still having trouble to apply/test these changes on Windows. Following up.

Reply:
@tovinkere : I thought there is no reason for me to get back to it, so I pushed a change that would remain you the principal owner of the method :)